_posts/2024/2024-02-06-about-caching-in-eo.md

```diff
@@ -0,0 +1,251 @@
+
+The goal of this blog is to research caching in frequently used build systems (`ccache`, `Maven`, `Gradle`)
+and to implement effective caching in [EO](https://github.com/objectionary/eo).
+
+<!--more-->
+
+## Build caching of existing build systems
+
+### ccache/sccache
+In compiled programming languages, building a project containing many source code files takes a long time.
+This time is spent on loading of libraries, preparing, optimizing, checking the code, and so on.
+To speed up the assembly of compiled languages, [ccache](https://ccache.dev) 
+and [sccache](https://github.com/mozilla/sccache) are used.
+Let's look at the assembly scheme using C++ as an example:
+
+<p align="center">
+  <img src="/images/defaultCPhase.svg">
+</p>
+
+1) First, preprocessor retrieves the source code files,
+which consist of both source files `.cpp` and header files `.h`.
+The result is a single file `.cpp` with human-readable code that the compiler will get.
+2) The compiler receives the edited code file `.cpp` and converts it into object file - `.obj`.
+At the compilation stage, parsing checks whether the code matches rules of a specific programming language.
+At the end, the compiler optimizes the resulting machine code and produces an object file. 
+To speed up compilation, different files of the same project might be compiled in parallel.
+3)  Then, the [Linker](https://en.wikipedia.org/wiki/Linker_(computing)) gets object files.
+The result of the linker is an executable `.exe` file.
+
+    
+`ccache` has hash algorithm, for the hashing of information to find cached files fast. 
+The [`ccache` hash](https://ccache.dev/manual/4.8.2.html#_common_hashed_information)
+includes information:
+* the file contents
+* the current directory of the file
+* the name of the compiler
+* the compiler’s size and modification time
+* extensions used by the compiler. 
+
+
+A compressed machine code file is placed in the cache using the received key.
+
+
+`ccache` has two main caching methods:
+1) `Direct mode` - hash is generated based on the source code. 
+`Direct mode` allows to build the program faster, since the preprocessor step is skipped.
+However, header files are not checked for changes, so the project may be built with not verified header files.
+2) `Preprocessor mode` - hash is generated based on the `.cpp` file received after the preprocessor step.
+`Preprocessor mode` is slower than `direct mode`, but the project is built with verified header files.
+
+`Sccache`, unlike `ccache`, allows you to store cached files not only locally, but also in a cloud data storage.
+And `sccache` supports a wider range of languages, while ccache focuses on caching C and C++ compiler.
+
+
+### Maven
+[Maven](https://maven.apache.org) automates and manages Java-project builds. 
+Building a project in `Maven` is completed in three
+[Maven LifeCycles](https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html),
```

Comment:
@Yanich96 Are you sure here? Yes, Maven definitely has three life cycles. However, to build a project you might use the only one.

Moreover, "Building a project in `Maven` is completed" has some particular [meanings](https://dictionary.cambridge.org/dictionary/english/complete). None of them are suitable here.